src/Npgsql/PreparedStatementManager.cs

```diff
@@ -98,161 +96,180 @@ internal PreparedStatementManager(NpgsqlConnector connector)
+            if (++pStatement.Usages < UsagesBeforePrepare)
             {
-                // We found a free or invalidated slot, exit the loop immediately
-                selectedIndex = i;
-                break;
+                // Statement still hasn't passed the usage threshold, no automatic preparation.
+                // Return null for unprepared execution.
+                pStatement.RefreshLastUsed();
+                return null;
             }
 
-            switch (slot.State)
+            // Bingo, we've just passed the usage threshold, statement should get prepared
+            LogMessages.AutoPreparingStatement(_commandLogger, sql, _connector.Id);
+
+            // Look for either an empty autoprepare slot, or the least recently used prepared statement which we'll replace it.
+            var oldestLastUsed = long.MaxValue;
+            var selectedIndex = -1;
+            for (var i = 0; i < AutoPrepared.Length; i++)
             {
-            case PreparedState.Prepared:
-                if (slot.LastUsed < oldestLastUsed)
+                var slot = AutoPrepared[i];
+
+                if (slot is null or { State: PreparedState.Invalidated })
                 {
+                    // We found a free or invalidated slot, exit the loop immediately
                     selectedIndex = i;
-                    oldestLastUsed = slot.LastUsed;
+                    break;
                 }
-                break;
 
-            case PreparedState.BeingPrepared:
-                // Slot has already been selected for preparation by an earlier statement in this batch. Skip it.
-                continue;
+                switch (slot.State)
+                {
+                case PreparedState.Prepared:
+                    if (slot.LastUsed < oldestLastUsed)
+                    {
+                        selectedIndex = i;
+                        oldestLastUsed = slot.LastUsed;
+                    }
+                    break;
 
-            default:
-                ThrowHelper.ThrowInvalidOperationException($"Invalid {nameof(PreparedState)} state {slot.State} encountered when scanning prepared statement slots");
-                return null;
+                case PreparedState.BeingPrepared:
+                    // Slot has already been selected for preparation by an earlier statement in this batch. Skip it.
+                    continue;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException($"Invalid {nameof(PreparedState)} state {slot.State} encountered when scanning prepared statement slots");
+                    return null;
+                }
             }
-        }
 
-        if (selectedIndex == -1)
-        {
-            // We're here if we couldn't find a free slot or a prepared statement to replace - this means all slots are taken by
-            // statements being prepared in this batch.
-            return null;
-        }
+            if (selectedIndex == -1)
```

Comment:
```suggestion
            if (selectedIndex < 0)
```
Also the CPU to fuse the compare and jump into only one actual executed instruction.
Kind of a micro-optimization...

Reply:
Tested, and the difference seems to be in using `test` instead of `cmp`, which saves us 1 byte out of 1325. That's definitely not a lot but the change itself is quite small and unobtrusive so why not...

Reply:
😄 
Sure that this won't move the needle much in almost any measure, but as you said: the change is small and helps a bit.
It helps mostly the CPU by just sending one (fused) command from the frontend to the backend. So less ports are occupied, etc. So if everywhere (where possible) the fused comparisons (e.g. `test` + `jmp`) would be used that could carray more effect.

Aside: Of course that should always be considered in the context (of code, environment, etc.) and one shouldn't artificially introduce unobvious code changes just to save a few bytes of instructions out of a huge stream of instruction.

Reply:
It's actually quite fascinating to me how small changes like this can actually improve perf (even if a little bit), so thank you for the insight!